ipi/engine/forcefields.py

```diff
@@ -885,3 +911,237 @@ def evaluate(self, r):
+
+    def start(self):
+        for ff in self.fflist:
+            ff.start()
+        super(FFCommittee, self).start()
+
+    def queue(self, atoms, cell, reqid=-1):
+
+        # launches requests for all of the committee FF objects
+        ffh = []
+        for ff in self.fflist:
+            ffh.append(ff.queue(atoms, cell, reqid))
+
+        # creates the request with the help of the base class,
+        # making sure it already contains a handle to the list of FF
+        # requests
+        req = super(FFCommittee, self).queue(
+            atoms, cell, reqid, template=dict(ff_handles=ffh)
+        )
+        return req
+
+    def check_finish(self, r):
+        """Checks if all sub-requests associated with a given
+        request are finished"""
+        for ff_r in r["ff_handles"]:
+            if ff_r["status"] != "Done":
+                return False
+        return True
+
+    def gather(self, r):
+        """ Collects results from all sub-requests, and assemble the committee of models. """
+
+        r["result"] = [0.0, np.zeros(len(r["pos"]), float), np.zeros((3, 3), float), ""]
+
+        # list of pointers to the forcefield requests. shallow copy so we can remove stuff
+        com_handles = r["ff_handles"].copy()
+        if self.baseline_name != "":
+            # looks for the baseline potential, store its value and drops it from the list
+            names = [ff.name for ff in self.fflist]
+
+            for i, ff_r in enumerate(com_handles):
+                if names[i] == self.baseline_name:
+                    baseline_pot = ff_r["result"][0]
+                    baseline_frc = ff_r["result"][1]
+                    baseline_vir = ff_r["result"][2]
+                    baseline_xtr = ff_r["result"][3]
+                    com_handles.pop(i)
+                    break
+
+        # Gathers the forcefield energetics and extras
+        pots = [ff_r["result"][0] for ff_r in com_handles]
+        frcs = [ff_r["result"][1] for ff_r in com_handles]
+        virs = [ff_r["result"][2] for ff_r in com_handles]
+        xtrs = [ff_r["result"][3] for ff_r in com_handles]
+
+        # Computes the mean energetics
+        mean_pot = np.mean(pots, axis=0)
+        mean_frc = np.mean(frcs, axis=0)
+        mean_vir = np.mean(virs, axis=0)
+
+        # Rescales the committee energetics so that their standard deviation corresponds to the error
+        rescaled_pots = np.asarray(
+            [mean_pot + self.alpha * (pot - mean_pot) for pot in pots]
+        )
+        rescaled_frcs = np.asarray(
+            [mean_frc + self.alpha * (frc - mean_frc) for frc in frcs]
+        )
+        rescaled_virs = np.asarray(
+            [mean_vir + self.alpha * (vir - mean_vir) for vir in virs]
+        )
+
+        # Calculates the error associated with the committee
+        var_pot = np.var(rescaled_pots, ddof=1)
+        std_pot = np.sqrt(var_pot)
+
+        if self.baseline_name != "":
```

Comment:
Maybe it is difficult for a reason I cannot immediately see, but the code below looks like something that would be much better encapsulated in one or two functions?

Reply:
it's not used anywhere else... how would you split it up in a way in which the function simplified the code more than by moving this rather lengthy block to a different place?

Reply:
ok, sometimes things like the uncertainty calls (there are at least 2 below right?) could be useful in many other parts of the code in the future, but sure. It is not hard to rewrite.